File: pset6/readability.py
from cs50 import get_string
from numpy import round


def main():

    text = get_string("Text: ")

    # Word, letter and sentence count
    words, letters, sentences = count_wls(text)

    # Avg letters & sentences per 100 words: count * (100 / words)
    L = letters * (100 / words)
    S = sentences * (100 / words)

    # Coleman-Liau index:
    cl = 0.0588 * L - 0.296 * S - 15.8
    cl = int(round(cl))

    if cl > 15:
        print("Grade 16+")
    elif cl < 1:
        print("Before Grade 1")
    else:
        print(f"Grade {cl}")


# Can all be done in one function, which is faster because it only loops through once


def count_wls(text):
    """Input string of text, output is a tuple of word count, letter count,
    sentence count"""

    words = 0
    letters = 0
    sentences = 0

    indicators = [".", "?", "!"]

    for c in text:

        # Word count
        if c == " ":
            words += 1

        # Letter count
        elif c.isalpha():
            letters += 1

        # Sentence count
        elif c in indicators:
            sentences += 1

    # Add one to word count to include the final word of the text
    words += 1

    return words, letters, sentences


# Alternatively, for more reusable code we can have the functions separate


def count_words(text):
    """Input a string of text, output is word count"""

    word_count = 0

    # Space indicates end of a word
    for c in text:
        if c == " ":
            word_count += 1

    # There is no space at the end of the text, so must + 1
    word_count += 1
    return word_count


def count_letters(text):
    """Input a string of text, output is the count of upper and lower
    case letters"""

    letter_count = 0

    # For each character, check it is a letter
    for c in text:
        if c.isalpha():
            letter_count += 1

    return letter_count


def count_sentences(text):
    """Input a string of text, output is a count of the sentences, defined
    as any sequence of chars ending in a period, exclamation mark or question mark"""

    indicators = [".", "?", "!"]
    sentence_count = 0

    for c in text:
        if c in indicators:
            sentence_count += 1

    return sentence_count


main()

File: readability.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <math.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);
float avg_per_100_words(int words, int count);

int main(void)
{
    string text = get_string("Text: ");

    //The Coleman-Liau index is index = 0.0588 * L - 0.296 * S - 15.8 where
    //L is avg letters per 100 words, and S is avg sentences per 100 words
    int words = count_words(text);
    float L = avg_per_100_words(words, count_letters(text));
    float S = avg_per_100_words(words, count_sentences(text));
    float index = 0.0588 * L - 0.296 * S - 15.8;

    //Round to determine the grade
    int grade = round(index);

    //Print the grade of the text
    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", grade);
    }
}


//Count the number of letters in a string
int count_letters(string text)
{
    int count = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        //Count lowercase
        if (text[i] >= 'a' && text[i] <= 'z')
        {
            count += 1;
        }
        //Count uppercase
        else if (text[i] >= 'A' && text[i] <= 'Z')
        {
            count += 1;
        }
    }

    return count;
}


//Count the number of words in a string
int count_words(string text)
{
    //Count the number of spaces in the text
    int count = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        //The ascii for space is 32
        if (text[i] == 32)
        {
            count += 1;
        }
    }
    //Word count is spaces + 1 (no space at end of text)
    return count + 1;
}


//Count sentences
int count_sentences(string text)
{
    int count = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        //We ignore the nuances and simply count anything that ends with . ! or ?
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            count += 1;
        }
    }

    return count;
}


//Average per hundred words
float avg_per_100_words(int words, int count)
{
    //Avg is count * (100 / words)
    return (float) count * (100 / (float) words);
}